src/rdkafka_metadata.c

```diff
@@ -1971,3 +1971,110 @@ rd_kafka_metadata_new_topic_with_partition_replicas_mock(int replication_factor,
+                /* TODO: How to fetch Proto? */
+                rd_kafka_broker_update(rk, RD_KAFKA_PROTO_PLAINTEXT,
+                                       &mdi->metadata.brokers[i], NULL);
+        }
+
+        for (i = 0; i < mdi->metadata.topic_cnt; i++) {
+                if (!RD_KAFKA_UUID_IS_ZERO(mdi->topics[i].topic_id)) {
+                        rkmce = rd_kafka_metadata_cache_find_by_id(
+                            rk, mdi->topics[i].topic_id, 1);
+                } else {
+                        rkmce = rd_kafka_metadata_cache_find(
+                            rk, mdi->metadata.topics[i].topic, 1);
+                }
+
+                if (!rkmce) {
+                        rd_kafka_log(
+                            rk, LOG_WARNING, "METADATA",
+                            "Metadata update: Topic %s not found in cache",
+                            mdi->metadata.topics[i].topic);
+                        continue;
+                }
+
+                for (j = 0; j < mdi->metadata.topics[i].partition_cnt; j++) {
+                        mdp     = &mdi->metadata.topics[i].partitions[j];
+                        mdp_int = &mdi->topics[i].partitions[j];
+
+                        if (mdp->id >= rkmce->rkmce_mtopic.partition_cnt) {
+                                rd_kafka_log(
+                                    rk, LOG_WARNING, "METADATA",
+                                    "Metadata update: Topic %s partition %d "
+                                    "not found in cache",
+                                    rkmce->rkmce_mtopic.topic, mdp->id);
+                                continue;
+                        }
+
+                        rkb = rd_kafka_broker_find_by_nodeid(rk, mdp->leader);
+                        if (!rkb) {
+                                rd_kafka_log(
+                                    rk, LOG_WARNING, "METADATA",
+                                    "Metadata update: Topic %s partition %d "
+                                    "leader %" PRId32 " not found in cache",
+                                    rkmce->rkmce_mtopic.topic, mdp->id,
+                                    mdp->leader);
+                                continue;
+                        }
+
+                        if (rkmce->rkmce_metadata_internal_topic
+                                .partitions[mdp->id]
+                                .leader_epoch >= mdp_int->leader_epoch) {
+                                rd_kafka_dbg(
+                                    rk, METADATA, "METADATA",
+                                    "Cache entry already has newer leader "
+                                    "epoch for topic %s partition %d: ",
+                                    rkmce->rkmce_mtopic.topic, mdp->id);
+                                continue;
+                        }
+                        cache_updated = rd_true;
+
+                        rkmce->rkmce_metadata_internal_topic.partitions[mdp->id]
```

Comment:
```suggestion
                        /* Need to acquire the write lock to avoid dirty reads
                         * from other threads acquiring read locks. */
                        rd_kafka_wrlock(rk);
                        rkmce->rkmce_metadata_internal_topic.partitions[mdp->id]
```

Reply:
Updated